### ISSUE TITLE
MinGW/Clang crash and premature deallocation using AngelScript in C++11 mode

### ISSUE THREAD

**Issue by Skrylar** [urho3d/Urho3D#708]
Environment:
MSYS
GCC 4.9.2
Clang 3.5.1
Windows 7

Situation:
Compiling under clang with C++11 enabled results in the AngelScript bindings being unusable.  After tracing around the debug and release builds under the same environment, it seems to be related with how strings are getting registered/allocated with AS.  Removing code that creates strings limps along a little further, so the errors seem to be focused around how strings/objects are being allocated.  At one point I also saw an error about an assertion volation where the refcount was off by one as well.

I'm not entirely sure what is causing this; it seems that calling functions works, as tracing through shows some methods do in fact work when the C++ and AS boundary is crossed.  It seems to be something related to constructors and reference counting, as though objects are being prematurely freed on their way out to AS  I did a third debug build under the native GCC and that particular build works, so it does appear to be compiler specific. Now the question is finding out exactly which part has to be dug around to fix _that_, since the rest of the codebase (including the samples written in C++) appears to work just fine in this setup.

**Comment by Skrylar** [comment]
I just tested the latest AngelScript in a test project with the same compiler, and it seems string support works fine there when using the built-in std::string library.  They seem to mention that some code gets inlined by clang which is unhelpful, so there's some workarounds in the angel code. I'll have to check and see if this is the same problem affecting current versions of Urho as well.

**Comment by weitjong** [comment]
I am not sure it's related to your issue or not. In my own Urho3D fork I have also attempted to use modern CPP standard and in doing so, one of the compiler (Clang if I recall it correctly) gave a number of compilation errors (or wanings, not very sure about that) in the AngelScript API binding code for assignment operator for a few classes. I am not in front of my workstation now so I could not double check that. I resolved them by using asMETHODPR macro instead of asMETHOD macro used by upstream Urho main repo. I also encountered similar compilation errors when I experimenting with Emscripten embind which uses "std=c++11" on a freshly cloned Urho3D repo.

So, a question for you. Did you encounter those compilation errors too when you use the new cpp standard? If yes, how did you resolve it?

**Comment by Skrylar** [comment]
I don't have compilation errors on my system, everything *compiles* fine.  When I try to run AngelScript code, it segfaults.  I did an experiment with taking the latest AS and testing it with clang, turns out std::string support works there just fine.  So now I'm trying to upgrade Urho to the latest AngelScript in the hopes that will solve the problem, which involves a bit of fiddling and fighting with deprecated functions and learning angelscript binding at the same time.

**Comment by Skrylar** [comment]
Looks like upgrading doesn't fix the issue on Clang.  I found out some of the errors I was poking through were the result of value assignments instead of handle assignments in the sample code, which must have been set to be allowed in the Urho configured build (but weren't in my test copy.)

Fiddling with this resulted in finding similar assertion errors with reference counting.  Guess its back to figuring out what voodoo is causing that.

**Comment by Skrylar** [comment]
As suspected, this seems to be an issue with compiler output.  Disassembling the GCC version shows newly created strings being returned in eax, my test environment with AngelScript and CLang also properly return through eax, however for some reason Urho3D::StringFactory's assembly is *not* returning the value through eax and is leading to stack corruption.

**Comment by GoogleBot42** [comment]
Maybe this is a compiler bug?  But weitjong isn't having this issue...  Are you sure you have a clean version of the latest Urho3D repo?  Maybe it has something to do with your clang version (what does the "TV" mean in "3.5.1TV").

**Comment by Skrylar** [comment]
The TV part was a typo, my bad.

It might be a compiler bug; looking at the disassembly shows that GCC is passing values through eax (which AngelScript expects) while Clang is using some other strange way of returning the String type.  Is he using the 3.7 pre-release or another version? I don't have a problem updating the compiler if it actually helps.

I found that if I used -fno-elide-constructors it would force clang to start pushing data as AngelScript expects again, however this lead to other crashes that are known Clang/LLVM bugs.

**Comment by weitjong** [comment]
I have time to double check on my fork and on my Emscripten embind experiment local repos. At the time I made the attempt to use the more modern cpp standard on my fork, I was still using Fedora prior to release 21 that I have now. My current system has Clang 3.5, so most probably I was using Clang 3.4 then. Same as my Emscripten embind experiment. It was done on Clang 3.4 (Emscripten 1.29) as Emscripten 1.30 with Clang 3.5 was just being released not long ago. Anyway, I suspect only Clang 3.4 gives the compilation errors for opAssign in the Angelscript binding code that I mentioned in my earlier comment. It appears Clang 3.5 does not produce them anymore. So, it is an improvement[?]

I just rebuilt Urho3DPlayer using Clang 3.5 on my current Fedora and I can use it to play NinjaSnowWar and Editor without running into any problem in my short tests. As you know, these two use AngelScript heavily. The only compilation error I got this time when building Urho3DPlayer was for [this line of code](https://github.com/urho3d/Urho3D/blob/master/Source/ThirdParty/JO/jo_jpeg.cpp#L265). You said Clang 3.5.1 on your Windows host system did not give you any errors at all, which still sounds strange to me. But since I have not used Clang on Windows before, probably I am not the best person to help you here.

**Comment by Skrylar** [comment]
@weitjong The version I see from git right now is using asMETHODPR as well, so that might be why some of the compile errors are gone?

I can check for newer versions of clang.  Digging around seems to indicate this problem is related to some Windows-specific conventions trying to be followed by clang which is messing things up.

**Comment by cadaver** [comment]
This may not necessarily be a bug, but a mismatch of calling convention assumptions.

It might be pleasant to eliminate the native calling convention from Urho's AngelScript use altogether, so that there would be no assembly code, no platform-specific mismatches, and AngelScript would work on Emscripten :)

As the generic bindings are a lot of repetitive code, this would practically require a bindings generation, which has been discussed / listed in the issues anyway

**Comment by Skrylar** [comment]
@cadaver Its a mismatch of Clang trying to use C++11 move semantics to make better code, but this ends up getting used by code oblivious to the change.  I used -fno-elide-constructors previously to confirm this is the case, but that caused other issues.  It seems there is now a Clang 3.6 out as a stable release, so I may need to upgrade my toolbox and see if the problem persists.

As for generic bindings, those would be nice to have on an optional basis.  I would suggest keeping the native bindings as an option for platforms where it works, as those are probably a little bit faster than going through the generic interface.  I'm not sure if the JIT system cares whether native bindings are used or not, though its quite nice when JITs can fully inline calls back in to C (as LuaJIT does with its FFI interface.)

**Comment by weitjong** [comment]
> This may not necessarily be a compiler bug, but a mismatch of calling convention assumptions.

I took the liberty to label the issue as bug. It is a bug, but fortunately not on our side. I thought "Windows Clang" is an oxymoron until today to find it actually exists. :smile:

> As the generic bindings are a lot of repetitive code, this would practically require a bindings generation, which has been discussed / listed in the issues anyway.

This is the reason I have abandoned my initial work to bind all our existing Angelscript API using generic calling convention. I lose interest after completing just the script objects portion. The work got "git stash" away and I don't think it would be popped out again anytime. Instead I have spent some of time to investigate the possibility of using "JS native calling convention" for AngelScript binding. This is also why I started experimenting with embind. Don't hold your breath for it though.

**Comment by Skrylar** [comment]
Doesn't AngelScript have a reflection system so you can check attributes or
custom meta-data on specific objects? Seems like a generator could be made
that just took a bunch of prototypes from an angelscript source file and
interrogated the metadata off of it.  A lot of the infrastructure is
already there it would seem (but this is maybe a different topic entirely.)

I'll check if 3.6 works today or not.

**Comment by cadaver** [comment]
It does, and that is what we use to generate the script API documentation. However this data is possible to inspect only *after* the engine C++ classes have been bound to the script engine. So the generation of script bindings is rather a question of parsing C++ headers, doxygen xml, or similar.

**Comment by Skrylar** [comment]
I just tested with a fresh Clang 3.6.  No dice.  After fishing around the assembly a bit more I noticed that the first call to String::operator+ has a correct right hand side, which is then corrupted as you step through the function.  It appears for some reason that AngelScript is receiving a copy of the String object which it then passes back to Urho, except between these two points the memory has been overwritten (something has marked that space as unallocated?)

When stepping through the ResizeString function with a memory breakpoint, it shows that the return string was allocated with the same memory as the original right hand side.  I'm not sure if this is still an assembly issue at this point, or a memory leak that seems to be exclusive to the builds I'm getting.

**Comment by Skrylar** [comment]
Ok, I fixed the problem (hooray.)  So the problem turned out that AngelScript has a special detection case for MinGW *and* GCC 4.7 and beyond.  This is because the calling conventions changed between 4.6 and 4.7, which does explain why using an older toolchain would have worked just fine.  Clang updated to support the newer 4.7 calling conventions.  Since AngelScript was not set up to detect Clang 3.6 as something that produced Mingw4.7 code, it was using the wrong calling conventions.

So, what I did to fix this in the short term is delete the ifdefs and force the correct calling conventions.  A longer-term solution is that I need to contact the AS developer to patch it upstream, and also include that modification to Urho's local copy.  Its the weekend now, but since a workable solution is found it shouldn't be too much longer before this is patchable.
